### PR TITLE
Context Menu Support!

### PR DESCRIPTION
**Pull request by VelvetThePanda**
This PR adds support for context-menus and responding to interactions created by such. 

This does **NOT** introduce a framework for creating/responding to context menus, though. 

## __**As of opening this PR, using context menus are in private beta.**__

### CODE REVIEW

File: DSharpPlus/Enums/Interaction/ApplicationCommandType.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public enum ApplicationCommandType
+    {
+        /// <summary>
+        /// This command is registered as a slash-command, aka "Chat Input".
+        /// </summary>
+        SlashCommand,
+        /// <summary>
+        /// This command is registered as a user context menu, and is applicable when interacting a user.
+        /// </summary>
+        UserConextMenu,
```
Comment (IDoEverything) [inline DSharpPlus/Enums/Interaction/ApplicationCommandType.cs:37]: `UserContextMenu`*

**Review by IDoEverything — CHANGES_REQUESTED** [review]

**Comment by Lulalaby** [comment]
Ah there it is

**Comment by Lulalaby** [comment]
@IDoEverything I think this should be On-Hold.

**Comment by VelvetThePanda** [comment]
#894 and #905 weren't put on hold

**Comment by Lulalaby** [comment]
They're released in the docs already, but i think context menu apps won't be released soon, am I right?

**Comment by Emzi0767** [comment]
There are changes pending to the API on Discord's side, and Velvet said this is also pending some extra work. I'm putting this on hold.

**Comment by VelvetThePanda** [comment]
"Pending some extra work" entailing me using the allegedly planned change to send the ApplicationCommandType in the interaction payload instead of relying on Data.Resolved.Users to determine the type of context menu.

**Comment by VelvetThePanda** [comment]
> They're released in the docs already, but i think context menu apps won't be released soon, am I right?

I forgot to mention, those were both pre-emptive PRs that were merged prior to api docs being merged, but we're here now so \*shrug\*

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 2]
```diff
@@ -1930,12 +1941,35 @@ internal async Task OnInteractionCreateAsync(ulong? guildId, ulong channelId, Tr
...
+                    interaction.Data.Resolved.Members?.TryGetValue(targetId, out targetMember);
+                    interaction.Data.Resolved.Users?.TryGetValue(targetId, out targetUser);
 
-                await this._interactionCreated.InvokeAsync(this, ea).ConfigureAwait(false);
+                    var ctea = new ContextMenuInteractionCreateEventArgs
+                    {
+                        Interaction = interaction,
+                        TargetUser = targetMember ?? targetUser,
+                        TargetMessage = targetMessage,
+                        Type = targetUser == null ? ApplicationCommandType.MessageContextMenu : ApplicationCommandType.UserContextMenu,
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:1960]: Wasn't this supposed to use some field discord sends that they didn't before

Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1960]: Do they send it now?

Comment (IDoEverything) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1960]: Dunno, doesn't really matter I suppose

**Review by IDoEverything — APPROVED** [review]

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 2 above)
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1960]: The type is sent now, but from what I can tell that's still an *OK* way of doing it. Not ideal, but you've merged this now so

**Pull request merged**